src/delta.c

```diff
@@ -38,86 +38,159 @@
 {
-	char *filename;
-	struct stat stat;
+	/* TODO: implement this without the indirection of creating a file... */
+	struct file f = {};
 
-	if (file->is_file == 0) {
-		return;
-	}
+	f.filename = filename;
+	f.use_xattrs = true;
 
-	if (file->deltapeer == NULL) {
-		return;
+	populate_file_struct(&f, filename);
+	if (compute_hash(&f, filename) != 0) {
+		return false;
 	}
 
-	if (file->deltapeer->is_file == 0) {
-		return;
-	}
+	hash_assign(&f.hash[0], hash);
+	return true;
+}
 
-	if (file->deltapeer->is_deleted) {
-		return;
+static void apply_one_delta(char *from_file, char *to_staged, char *delta_file, char *to_hash)
+{
+	int ret;
+
+	char *prefixed_from_file;
+	string_or_die(&prefixed_from_file, "%s/%s", path_prefix, from_file);
+
+	ret = apply_bsdiff_delta(prefixed_from_file, to_staged, delta_file);
+	if (ret) {
+		goto end;
 	}
 
-	/* check if the full file is there already, because if it is, don't do the delta */
-	string_or_die(&filename, "%s/staged/%s", state_dir, file->hash);
-	if (lstat(filename, &stat) == 0) {
-		free_string(&filename);
-		return;
+	xattrs_copy(prefixed_from_file, to_staged);
+
+	char hash[SWUPD_HASH_LEN];
+	if (!compute_hash_from_file(to_staged, &hash[0])) {
+		fprintf(stderr, "Couldn't use delta file %s: hash calculation failed\n", delta_file);
+		goto remove_and_fail;
 	}
-	do_delta(filename, file);
-	free_string(&filename);
+	if (!hash_equal(hash, to_hash)) {
+		fprintf(stderr, "Couldn't use delta file %s: application resulted in wrong hash\n", delta_file);
+		goto remove_and_fail;
+	}
+
+	goto end;
+
+remove_and_fail:
+	(void)remove(to_staged);
+end:
+	free_string(&prefixed_from_file);
 }
 
-static void do_delta(char *filename, struct file *file)
+/* Check if the delta filename is well-formed, if so return true and fill the from/to
+ * buffers with the corresponding hashes. Return false otherwise. */
+static bool check_delta_filename(const char *delta_name, char *from, char *to)
 {
-	char *origin;
-	char *dir, *base, *tmp = NULL, *tmp2 = NULL;
-	char *deltafile = NULL;
-	int ret;
-	struct stat sb;
+	/* Delta files have the form [FROM_VERSION]-[TO_VERSION]-[FROM_HASH]-[TO_HASH]. */
+	const char *s = delta_name;
+
+	/* Ignore versions, deltas will be used based on their hashes only. */
+	for (int i = 0; i < 2; i++) {
+		s = strchr(s, '-');
+		if (!s) {
+			return false;
+		}
+		s++;
+	}
 
-	string_or_die(&deltafile, "%s/delta/%i-%i-%s-%s", state_dir,
-		      file->deltapeer->last_change, file->last_change, file->deltapeer->hash, file->hash);
+	/* Note: SWUPD_HASH_LEN accounts for the NUL-terminator after the hash. */
+	const size_t hash_len = SWUPD_HASH_LEN - 1;
+	if (strlen(s) != (hash_len * 2 + 1)) {
+		return false;
+	}
 
-	ret = stat(deltafile, &sb);
-	if (ret != 0) {
-		free_string(&deltafile);
-		return;
+	if (s[hash_len] != '-') {
+		return false;
 	}
 
-	tmp = strdup(file->deltapeer->filename);
-	tmp2 = strdup(file->deltapeer->filename);
+	memcpy(from, s, hash_len);
```

Comment:
Probably ought to be using hash_assign here rather than memcpy.

Reply:
Done.